### PR TITLE
feat: `DelayedOracle` store invalid feed

### PR DESCRIPTION
**Pull request by 0xJabberwock**
Closes HAI-304

### CODE REVIEW

**Review by 0xJabberwock — COMMENTED** [review]
Code comments are much appreciated! ✅

File: src/contracts/oracles/DelayedOracle.sol [hunk 1]
```diff
@@ -56,19 +56,35 @@ contract DelayedOracle is IBaseOracle, IDelayedOracle {
...
   function updateResult() external returns (bool _success) {
-    // Check if the delay passed
-    if (!_delayHasElapsed()) revert DelayedOracle_DelayHasNotElapsed();
+    // Check if the delay to set the new feed passed
+    if (!_delayHasElapsed()) {
+      // If it hasn't passed, check if the upcoming feed is valid
+      // in the case that it is not valid we check if we can fetch a new valid feed to replace it.
+      if (!_nextFeed.isValid) {
+        // Read the price from the median
+        (uint256 _priceFeedValue, bool _hasValidValue) = _getPriceSourceResult();
```
Comment (wei3erHase) [inline src/contracts/oracles/DelayedOracle.sol]: hmmm feels weird that there are 2 different lines (65 and 78) that do the same, please refactor this logic 🙏

**Review by wei3erHase — CHANGES_REQUESTED** [review]
please simplify logic 🙏

File: src/contracts/oracles/DelayedOracle.sol (hunk 1 above)
Comment (xBA5ED) [reply src/contracts/oracles/DelayedOracle.sol]: True, it was done to save some gas by not doing the `_getPriceSourceResult()` call in the case that the revert case was hit. I'll make it so the call is done beforehand

File: src/contracts/oracles/DelayedOracle.sol [hunk 2]
```diff
@@ -56,19 +56,33 @@ contract DelayedOracle is IBaseOracle, IDelayedOracle {
...
+    if (!_delayHasElapsed()) {
+      // If it hasn't passed, check if the upcoming feed is valid
+      // in the case that it is not valid we check if we can fetch a new valid feed to replace it.
+      if (!_nextFeed.isValid) {
+        // Check if the newly fetched feed is valid
+        if (_hasValidValue) {
+          _nextFeed = Feed(_priceFeedValue, _hasValidValue);
+        }
+        // Store the new next Feed
+        return _hasValidValue;
```
Comment (wei3erHase) [inline src/contracts/oracles/DelayedOracle.sol]: hmmmmmmmmmmm
i think we need to update the `lastUpdateTime`
the feed must be available 1hr (or delay time) after the valid feed was fetched (not after the 1st invalid one)

**Review by wei3erHase — CHANGES_REQUESTED** [review]
please update the `lastUpdateTime` also when re-fetching an invalid feed 🙏

File: src/contracts/oracles/DelayedOracle.sol [hunk 3]
```diff
@@ -44,7 +44,7 @@ contract DelayedOracle is IBaseOracle, IDelayedOracle {
 
     (uint256 _priceFeedValue, bool _hasValidValue) = _getPriceSourceResult();
     if (_hasValidValue) {
-      _nextFeed = Feed(_priceFeedValue, true);
+      _nextFeed = Feed(_priceFeedValue, _hasValidValue);
```
Comment (wei3erHase) [inline src/contracts/oracles/DelayedOracle.sol:47]: GREAT catch.

File: src/contracts/oracles/DelayedOracle.sol [hunk 4]
```diff
@@ -87,7 +103,7 @@ contract DelayedOracle is IBaseOracle, IDelayedOracle {
 
   /// @inheritdoc IDelayedOracle
   function shouldUpdate() external view returns (bool _ok) {
-    return _delayHasElapsed();
+    return _delayHasElapsed() || !_nextFeed.isValid;
```
Comment (wei3erHase) [inline src/contracts/oracles/DelayedOracle.sol:106]: love you pal 😍

**Review by wei3erHase — APPROVED** [review]
chapeau! 🎩

**Pull request merged**